csrc/preseg_passes/translate_no_reduction_matmul_to_mul_squeeze.cpp

```diff
@@ -0,0 +1,230 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <ir/utils.h>
+#include <iter_visitor.h>
+#include <logical_domain_map.h>
+#include <ops/all_ops.h>
+#include <preseg_passes/translate_no_reduction_matmul_to_mul_squeeze.h>
+
+#include <vector>
+
+namespace nvfuser::preseg_passes {
+
+namespace {
+
+// Translation algorithm overview:
+//
+// Step 1: Inspection. Traverses the given fusion and looks for a
+// sequence of MatmulOps with K=1.
+//
+// Step 2: Apply the translation.
+class NoReductionMatmulToMulSqueezeTranslator {
+ public:
+  NoReductionMatmulToMulSqueezeTranslator(Fusion* fusion) : fusion_(fusion) {}
+
+  void run() {
+    inspect();
+    translate();
+  }
+
+ private:
+  // Traverse through the fusion and gather all MatmulOps with K=1.
+  void inspect() {
+    const auto exprs = fusion_->exprs();
+
+    for (auto matmul : ir_utils::filterByType<MatmulOp>(exprs)) {
+      // Find the K dimension
+      auto in_a = matmul->inA();
+      auto in_b = matmul->inB();
+
+      auto domain_a = TensorDomain::noReductions(in_a->getLogicalDomain());
+      auto domain_b = TensorDomain::noReductions(in_b->getLogicalDomain());
+
+      auto ndims_a = domain_a.size();
+      auto ndims_b = domain_b.size();
+
+      if (ndims_a == 1 && ndims_b == 1) {
+        // Dot product. Should this be translated to a reduction as
+        // well?
+        continue;
+      }
+
+      IterDomain* k_id_a = domain_a.back();
+
+      if (!k_id_a->isBroadcast() || k_id_a->hasExpandedExtent()) {
+        continue;
+      }
+
+      // There should be no case where this extent is not one.
+      NVF_ERROR(
+          k_id_a->extent()->isOneInt(),
+          "Unexpected broadcast dimension: ",
+          k_id_a->toString());
+
```

Comment:
We should also add a check that the K axis is broadcast in both A and B.

Reply:
I don't think that's necessary. The broadcasting semantics doesn't apply to the matrix dimensions, IIUC.

Reply:
I was referring to this check: https://github.com/NVIDIA/Fuser/blob/42058034b16fb603dd3f518873848741789ddde2/csrc/ops/composite.cpp#L411-L414. Since it is already present in the construction of `MatmulOp` we can skip it here.